### ISSUE TITLE
Support ability to pass in placeholder values when deploying

### ISSUE THREAD

**Issue by MarkRx** [xebialabs-community/xlr-xldeploy-plugin#32]
I need the ability to dynamically pass in placeholder values from XLR to XLD. It is possible by simply modifying the deployment XML post but at present the only way to do so is hacky (#31). I can't just use dictionaries since they change with every deployment. Using dictionaries would require one per each application (or else concurrent deployments become a problem) which would mean 1000's of them which isn't maintainable.

**Comment by jdewinne** [comment]
Hi Mark,

Can you have a look if [v2.7.0](https://github.com/xebialabs-community/xlr-xldeploy-plugin/releases/tag/v2.7.0) fits your need. I've added the overrideDeployedProps.

**Comment by MarkRx** [comment]
@px-djozis what do you think? XPath seems necessary since placeholders store the key in the attributes section. 

For usability would it make sense to use findall so that the XPath expression could modify all matches at once? It would make things easier. If someone wanted to just modify one thing then they can just use the exact match for the XPath expression.

Is it named properly? It was intended for deployed properties but it looks like it can be used to modify the entire deployment XML (https://docs.xebialabs.com/generated/xl-deploy/5.1.x/rest-api/com.xebialabs.deployit.engine.api.dto.Deployment.html).

**Comment by px-djozis** [comment]
I don't personally find xpath an intuitive way to set placeholders. I think having a way to specify a list of deployable id matches supporting \*\*/\* style wildcarding and their placeholder substitutions to make would be both flexible and intuitive, since you asked for my input.

**Comment by MarkRx** [comment]
David and I talked about it. We don't actually need the power of xpath - I was simply trying to use it to achieve what I wanted in #31. We think a more user friendly impl would be this:

A dictionary of id to a dictionary of key, value pairs. The property would be "overridePlaceholders" or something like that.

> { "id1" : { "key": "value" }, { "key2" : "value2" }, "id2": ... }

"id" would correspond to the id in XLD. It should be able to use the standard "ant style" path wildcarding:

> - ** matches zero or more directories
> - *matches zero or more characters
> - ? matches one character

Examples:

Set "mykey" on all placeholders to "myvalue":

> { "**/*": {"mykey" : "myvalue" } }

Set "mykey" on "Infrastructure/Folder1/Foo/Bar" to "myvalue": 

> { "Infrastructure/Folder1/Foo/Bar": {"mykey" : "myvalue" } }

Set "mykey" on all deployeds named "Bar" to "myvalue": 

> { "**/Bar": {"mykey" : "myvalue" } }

Set "mykey" on all deployeds ending in "ar" to "myvalue": 

> { "**/*ar": {"mykey" : "myvalue" } }

Set "mykey" and "mykey2" to "myvalue" and "myvalue2" respectively on all placeholders:

> { "**/*": {"mykey" : "myvalue", "mykey2" : "myvalue2" } }

**Comment by jdewinne** [comment]
Hi Mark,

That would be limiting this feature into placeholders only. Which I don't like, as for example you might also want to change non placeholder deployed properties. So for the time being I'll stick with xpath.
Maybe we should go back to the beginning, and ask why you want to change placeholders/deployed properties during deployment (as that is some anti pattern already).

Regards,
Joris

**Comment by MarkRx** [comment]
@jdewinne could you at least please change the find to a findall and loop over all xpath matches? This would make things much easier. That would allow setting all dictionary placeholders of a certain key at once.

The reason we are using placeholder replacement because the placeholders represent the state of an application in a blue/green deployment. Storing the state in dictionaries becomes complicated because of concurrency issues and annoying application restrictions that would have to be set.

**Comment by MarkRx** [comment]
This would do it:

```
def override_deployed_properties(deployment_xml, deployed_properties):
    root = ET.fromstring(deployment_xml)
    if deployed_properties:
        deployeds_properties_dict = dict(ast.literal_eval(deployed_properties))
        for key in deployeds_properties_dict:
            for pkey_xml in root.findall(key):
                pkey_xml.text = deployeds_properties_dict[key]
    return ET.tostring(root)
```

Hence I can do something like this:

> '''[('deployeds/openshift.ResourceModule/placeholders/entry[@key="openshift.stage.color"]', 'blue'),
> ('deployeds/openshift.ResourceModule/placeholders/entry[@key="openshift.active.color"]', 'green')]'''

Instead of manually listing out all id's. The id's approach is also problematic because it requires knowledge of that whose will be and what deployables will actually end up creating deployeds.

**Comment by jdewinne** [comment]
Released [v2.8.0](https://github.com/xebialabs-community/xlr-xldeploy-plugin/releases/tag/v2.8.0) with findall in it.

**Comment by MarkRx** [comment]
A coworker working on this has come across an issue with the documentation that also demonstrates a limitation of the plugin:

> overrideDeployedProps (Map contain xpath expression to override deployed properties.) e.g.: {'deployeds/openshift.ResourceModule[@id="Infrastructure/Server/Openshift/project/myapp"]/placeholders/entry[@key="openshift.placeholder"]', 'value'}

The xpath @id expression must be an exact match. It will not match a subet and appears it does not work with wild carding. xpath does have a contains() function but the python ET implementation does not support it.

As a result either 
1. all matching deployeds must have placeholders set at the same time or 
2. each deployed must be explicitly configured. 

The problem with the first is it may not be flexible enough. The problem with the second is you won't know the deployed paths until the deployment is created (you could make a guess based on the application but without replicating tag processing an restrictions you wouldn't be 100% sure). The deployment is created by this plugin but there is no way to get that information in-between the initialization of the deployment and the filling in of the deployed properties. Hence deployeds cannot be modified based on information that is dependent on the deployed first being initialized.
